src/ballet/http/fd_hcache.h

```diff
@@ -0,0 +1,193 @@
+   hcache.  Assumes hcache is a current local join.
+
+   If there is not sufficient space in the data region to append the
+   formatted string, existing data will be overwritten.  Any connections
+   in the HTTP server which are still using the data which is being
+   overwritten will be closed, as they cannot keep up.
+
+   Once the full desired string has been appended into the hcache it can
+   be sent to all connected WebSocket clients of the HTTP server using
+   fd_hcache_snap_ws_broadcast.  This will snap off the message printed
+   so far and reset the hcache to start preparing the next message.
+
+   Printing is not error-free, it is assumed that the format string is
+   valid but the entire data region may not be large enough to hold the
+   printed string.  In that case, the hcache is marked as being in an
+   error state internally.  The next call to fd_hcache_snap_ws_broadcast
+   will fail and the error state will be cleared. */
+
+void
+fd_hcache_printf( fd_hcache_t *      hcache,
+                  char const *       fmt,
+                  ... );
+
+/* fd_hcache_memcpy appends the data into the current hcache frame.
+   Assumes hcache is a current local join.
+
+   If there is not sufficient space in the data region to append the
+   data, existing data will be overwritten.  Any connections in the HTTP
+   server which are still using the data which is being overwritten will
+   be closed, as they cannot keep up.
+
+   Once the full desired data has been appended into the hcache it can
+   be sent to all connected WebSocket clients of the HTTP server using
+   fd_hcache_snap_ws_broadcast.  This will snap off the message printed
+   so far and reset the hcache to start preparing the next message.
+
+   Copying is not error-free, it is assumed that the data is valid but
+   the entire data region may not be large enough to hold the copied
+   data.  In that case, the hcache is marked as being in an error state
+   internally.  The next call to fd_hcache_snap_ws_broadcast will fail
+   and the error state will be cleared. */
+
+void
+fd_hcache_memcpy( fd_hcache_t * hcache,
+                  uchar const * data,
+                  ulong         data_len );
+
+/* fd_hcache_snap_response takes the current contents of the hcache and
+   returns it as data which can be sent as an HTTP response body.  The
+   hcache is reset to start preparing the next message.  Assumes hcache
+   is a current local join.
+
+   The caller should return the body data as a response body to an HTTP
+   request immediately, as this is how the hcache manages the lifetime
+   of the buffer.  If the buffer is not being used to send a response
+   to any connected HTTP client, it could be overwritten by the next
+   snap and become no longer valid.
+   
+   Returns NULL on failure if the hcache is in an error state because
+   the message being snapped exceeds the entire hcache data size. */
+
```

Comment:
I also need a fd_hcache_reset method which sets snap_len to zero. This is useful if we need to switch to an error response mid-stream